CB-11183 Appium tests: Added image verification

alsorokin opened the pull request:
https://issues.apache.org/jira/browse/CB-11183

nikhilkh commented:
@riknoll  to help review

riknoll commented:
It would be nice to refactor the specs a bit, because they seem to be 90% shared code. This looks pretty good! Reading the file and checking the dimensions covers a lot of the bugs for camera.

Here are some other things that can be verified:

* Make sure that PNG files and JPEG files are actually encoded/named correctly (check the headers). We've had bugs in the past where we return a JPEG file with a .png file name
* Run specs with quality set to 100 in addition to 50. We often treat a quality of 100 as a special case on iOS and Android

It would also be nice to verify the type of URI we get back for the destination type. Unfortunately, that is horribly inconsistent right now on Android for NATIVE_URI and FILE_URI. It is possible to get back a `file:` URI, a `content:` URI, or a vanilla path depending on what camera options you pass. Can we try and verify it on iOS at least?

I think that covers most of the camera options that can be tested.

alsorokin commented:
@riknoll I've updated the PR.
* De-duplicated specs code
* Implemented image type verification
* Added a couple of specs with quality 100 (adding it to combined specs would double their number)
* Implemented URI type verification for iOS

riknoll commented on appium-tests/helpers/cameraHelper.js:
```diff
@@ -103,11 +109,185 @@ module.exports.getPicture = function (opts, pid) {
+    // skip image type check if it's unmodified on Android:
+    // https://github.com/apache/cordova-plugin-camera/#android-quirks-1
+    var skipFileTypeCheck = isAndroid &&
+        (!options.quality || options.quality === 100) &&
```
Why check for `!options.quality`? The default quality is 80 on Android (though we document it as 50!)

alsorokin replied:
Because this options is what we pass from the tests. Some tests do not specify quality.

jcesarmobile replied:
@riknoll the truth is quality default is 50 for all platforms as if no quality option is passed, Camera.js set it to 50 (https://github.com/apache/cordova-plugin-camera/blob/master/www/Camera.js#L145). Even if the java code sets the default to 80 it will be 50 (I have tested). Anyway, I'm going to open an issue and set it 50 in the java code too.

alsorokin replied:
@riknoll actually if I remove that check, some specs with default quality are failing. For example, this options: `sourceType: PHOTOLIBRARY, destinationType: FILE_URI, encodingType: PNG, allowEdit: false, correctOrientation: false` return JPEG file.

Is this bug, so I should leave this check here?

riknoll replied:
That does sound like a bug! Only time you should get a JPEG with PNG set is the quirk that you linked (when quality=100)

riknoll replied:
Thanks @jcesarmobile! We should probably just remove the defaults being set in the Java code altogether to avoid confusion in the future

nikhilkh replied:
Should the default be 50? I believe the intent was that it will use lesser memory and hence, default of 50 is a good thing - but I did some profiling and setting the quality to anything but 100 on Android actually causes more memory consumption because the image has to be loaded as a bitmap and re-compressed. Setting appropriate target height & width along with FILE_URI is the correct way to reduce the memory requirements. I propose that the default be changed to 100.

alsorokin replied:
Default quality 100 sounds like a good idea, especially if it'll help reduce memory usage.
It is also intuitive: I was expecting it to be 100.

riknoll commented:
Awesome! LGTM other than the one comment

cordova-qa commented:
Cordova CI Build has completed successfully. 

**Dashboard** - http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3/

**Results**
1. [Windows 8.1 Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=windows-8.1-store/) - [Console Output]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=windows-8.1-store/console) - [Test Results]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=windows-8.1-store/testReport/) - [Device Logs]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=windows-8.1-store/artifact/)
2. [Windows 10  Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=windows-10-store/) - [Console Output]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=windows-10-store/console) - [Test Results]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=windows-10-store/testReport/) - [Device Logs]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=windows-10-store/artifact/)
3. [Windows 8.1 Phone]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=windows-8.1-phone/) - [Console Output]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=windows-8.1-phone/console) - [Test Results]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=windows-8.1-phone/testReport/) - [Device Logs]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=windows-8.1-phone/artifact/)
4. [iOS]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=mac-slave,platformName=ios/) - [Console Output]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=mac-slave,platformName=ios/console) - [Test Results]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=mac-slave,platformName=ios/testReport/) - [Build Artifacts]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=mac-slave,platformName=ios/artifact/)
5. [Android Mac]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=mac-slave,platformName=android/) - [Console Output]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=mac-slave,platformName=android/console) - [Test Results]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=mac-slave,platformName=android/testReport/) - [Build Artifacts]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=mac-slave,platformName=android/artifact/)
6. [Android Windows]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=android/) - [Console Output]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=android/console) - [Test Results]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=android/testReport/) - [Device Logs]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera/3//label=windows-slave,platformName=android/artifact/)

cordova-qa commented:
Cordova CI Build has completed successfully. 

**Commit**     - [Link](https://github.com/apache/cordova-plugin-camera/pull/209/commits/278b5277027ba62d621cbc2dea10079abb2ef0c9)
**Dashboard** - [Link](http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12/)

| Builder Name  | Console Output | Test Report | Device Logs  |
|     :---:     |     :---:      |   :---:     |     :---:    |
| [Windows 8.1 Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=windows-slave,platformName=windows-8.1-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=windows-slave,platformName=windows-8.1-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=windows-slave,platformName=windows-8.1-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=windows-slave,platformName=windows-8.1-store/artifact/) |
| [Windows 10  Store]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=windows-slave,platformName=windows-10-store/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=windows-slave,platformName=windows-10-store/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=windows-slave,platformName=windows-10-store/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=windows-slave,platformName=windows-10-store/artifact/) |
| [Windows 8.1 Phone]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=windows-slave,platformName=windows-8.1-phone/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=windows-slave,platformName=windows-8.1-phone/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=windows-slave,platformName=windows-8.1-phone/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=windows-slave,platformName=windows-8.1-phone/artifact/) |
| [iOS]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=mac-slave,platformName=ios/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=mac-slave,platformName=ios/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=mac-slave,platformName=ios/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=mac-slave,platformName=ios/artifact/) |
| [Android Mac]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=mac-slave,platformName=android/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=mac-slave,platformName=android/console) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=mac-slave,platformName=android/testReport/) | [Link]( http://cordova-ci.cloudapp.net:8080/job/cordova-plugin-camera-pr/12//label=mac-slave,platformName=android/artifact/) |
